pyomo/contrib/doe/DoE_package.py

```diff
@@ -0,0 +1,1812 @@
+        plt.rc('xtick', labelsize=font_tick)
+        plt.rc('ytick', labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {'mathtext.default': 'regular'}
+        plt.rcParams.update(params)
+        ax.set_yticks(range(len(yLabel)))
+        ax.set_yticklabels(yLabel)
+        ax.set_ylabel(ylabel_text)
+        ax.set_xticks(range(len(xLabel)))
+        ax.set_xticklabels(xLabel)
+        ax.set_xlabel(xlabel_text)
+        im = ax.imshow(hes_a.T, cmap=plt.cm.hot_r)
+        ba = plt.colorbar(im)
+        ba.set_label('log10(trace(FIM))')
+        plt.title(title_text + ' - A optimality')
+        plt.show()
+
+        fig = plt.figure()
+        plt.rc('axes', titlesize=font_axes)
+        plt.rc('axes', labelsize=font_axes)
+        plt.rc('xtick', labelsize=font_tick)
+        plt.rc('ytick', labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {'mathtext.default': 'regular'}
+        plt.rcParams.update(params)
+        ax.set_yticks(range(len(yLabel)))
+        ax.set_yticklabels(yLabel)
+        ax.set_ylabel(ylabel_text)
+        ax.set_xticks(range(len(xLabel)))
+        ax.set_xticklabels(xLabel)
+        ax.set_xlabel(xlabel_text)
+        im = ax.imshow(hes_d.T, cmap=plt.cm.hot_r)
+        ba = plt.colorbar(im)
+        ba.set_label('log10(det(FIM))')
+        plt.title(title_text + ' - D optimality')
+        plt.show()
+
+        fig = plt.figure()
+        plt.rc('axes', titlesize=font_axes)
+        plt.rc('axes', labelsize=font_axes)
+        plt.rc('xtick', labelsize=font_tick)
+        plt.rc('ytick', labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {'mathtext.default': 'regular'}
+        plt.rcParams.update(params)
+        ax.set_yticks(range(len(yLabel)))
+        ax.set_yticklabels(yLabel)
+        ax.set_ylabel(ylabel_text)
+        ax.set_xticks(range(len(xLabel)))
+        ax.set_xticklabels(xLabel)
+        ax.set_xlabel(xlabel_text)
+        im = ax.imshow(hes_e.T, cmap=plt.cm.hot_r)
+        ba = plt.colorbar(im)
+        ba.set_label('log10(minimal eig(FIM))')
+        plt.title(title_text + ' - E optimality')
+        plt.show()
+
+        fig = plt.figure()
+        plt.rc('axes', titlesize=font_axes)
+        plt.rc('axes', labelsize=font_axes)
+        plt.rc('xtick', labelsize=font_tick)
+        plt.rc('ytick', labelsize=font_tick)
+        ax = fig.add_subplot(111)
+        params = {'mathtext.default': 'regular'}
+        plt.rcParams.update(params)
+        ax.set_yticks(range(len(yLabel)))
+        ax.set_yticklabels(yLabel)
+        ax.set_ylabel(ylabel_text)
+        ax.set_xticks(range(len(xLabel)))
+        ax.set_xticklabels(xLabel)
+        ax.set_xlabel(xlabel_text)
+        im = ax.imshow(hes_e2.T, cmap=plt.cm.hot_r)
+        ba = plt.colorbar(im)
+        ba.set_label('log10(cond(FIM))')
+        plt.title(title_text + ' - Modified E-optimality')
+        plt.show()
+
+
+def permutations(lst, r=None):
```

Comment:
Can we replace this with https://docs.python.org/3/library/itertools.html#itertools.permutations?

One problem is that code from stackoverflow may not be compatible with the BSD3 license (used by Pyomo).

Reply:
Yes. Replaced.